### PR TITLE
Refacftored OAuth2

### PR DESCRIPTION
**Pull request by ritikk112**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-library/issues/7237

**Refactored SSL context initialization**: Improved logic for handling optional SSL keys and certificates.
**Enhanced exception handling**: Streamlined error catching for IOException and InterruptedException.
**Optimized code structure**: Removed redundant checks and simplified nested if-else statements.
**Improved readability**: Cleaned up and made the code more readable across methods.
## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

**Comment by MohamedSabthar** [comment]
Hi @ritikk112, 

Are you working on this issue: https://github.com/ballerina-platform/ballerina-library/issues/7237? If so, could you refactor the Ballerina codebase? The current Java code appears to follow best practices, so changes to it are not necessary.

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 1]
```diff
@@ -70,16 +71,16 @@ public static Object doHttpRequest(BString url, BMap<BString, Object> clientConf
             headersList.add(entry.getValue().getValue());
         }
 
-        BMap<BString, ?> customHeaders = getBMapValueIfPresent(clientConfig, OAuth2Constants.CUSTOM_HEADERS);
-        if (customHeaders != null) {
+        Optional<BMap<BString, ?>> customHeaders = getBMapValueIfPresent(clientConfig, OAuth2Constants.CUSTOM_HEADERS);
```
Comment (MohamedSabthar) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:74]: Replacing null checks with Optional types doesn't add any value here, as we are following an imperative style of coding. The previous code block looks fine, so let's keep the original code.

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 2]
```diff
@@ -70,16 +71,16 @@ public static Object doHttpRequest(BString url, BMap<BString, Object> clientConf
...
             for (Map.Entry<BString, ?> entry : customHeaders.entrySet()) {
                 headersList.add(entry.getKey().getValue());
                 headersList.add(((BString) entry.getValue()).getValue());
             }
-        }
+        });
 
         String httpVersion = getBStringValueIfPresent(clientConfig, OAuth2Constants.HTTP_VERSION).getValue();
-        BMap<BString, ?> secureSocket = getBMapValueIfPresent(clientConfig, OAuth2Constants.SECURE_SOCKET);
+        Optional<BMap<BString, ?>> secureSocket = getBMapValueIfPresent(clientConfig, OAuth2Constants.SECURE_SOCKET);
```
Comment (MohamedSabthar) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:83]: ditto

Comment (MohamedSabthar) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:83]: check other places.

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 3]
```diff
@@ -301,6 +299,9 @@ private static Object callEndpoint(HttpClient client, HttpRequest request) {
             return createError("Failed to get a success response from the endpoint. Response code: '" +
                                        response.statusCode() + "'. Response body: '" + response.body() + "'");
         } catch (IOException | InterruptedException e) {
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt(); // Restore interrupted status
+            }
```
Comment (MohamedSabthar) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:304]: Here we have already handled the InterruptedException by throwing an error to the Ballerina side via the `createError` method. It's not necessary to restore the status here. Is there a any particular reason to introduce this change ?

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java (hunk 1 above)
Comment (ritikk112) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:74]: I introduced Optional checks to improve code readability as there were a lot of if else statements prior to changes and made the code less streamlined, should I revert back to if checks for null values?

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java (hunk 3 above)
Comment (ritikk112) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:304]: I added this just for additional security measures and to handle errors more gracefully

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java (hunk 1 above)
Comment (MohamedSabthar) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:74]: Let's remove.

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java (hunk 3 above)
Comment (MohamedSabthar) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:304]: Let's remove this as there's no need to propagate this interruption.

**Comment by ritikk112** [comment]
HI @MohamedSabthar made the appropriate changes, can you do a quick review and let me know if any other changes are to made.

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 4]
```diff
@@ -71,7 +71,7 @@ public static Object doHttpRequest(BString url, BMap<BString, Object> clientConf
         }
 
         BMap<BString, ?> customHeaders = getBMapValueIfPresent(clientConfig, OAuth2Constants.CUSTOM_HEADERS);
-        if (customHeaders != null) {
+        if(customHeaders != null){
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:74]: Why are these spaces removed?

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 5]
```diff
@@ -104,20 +103,18 @@ public static Object doHttpRequest(BString url, BMap<BString, Object> clientConf
             } catch (Exception e) {
                 return createError("Failed to init SSL context. " + e.getMessage());
             }
-        }
+        }
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:106]: Remove additional spaces. You can configure your IDE to remove the trailing spaces when saving a file.

```suggestion
        }
```

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 6]
```diff
@@ -104,20 +103,18 @@ public static Object doHttpRequest(BString url, BMap<BString, Object> clientConf
...
+        }        
         HttpClient client = buildHttpClient(httpVersion);
         return callEndpoint(client, request);
     }
 
     private static URI buildUri(String url, BMap<BString, ?> secureSocket) throws IllegalArgumentException {
         String[] urlParts = url.split(OAuth2Constants.SCHEME_SEPARATOR, 2);
         if (urlParts.length == 1) {
-            urlParts = secureSocket != null ? new String[]{OAuth2Constants.HTTPS_SCHEME, urlParts[0]} :
+            urlParts = (secureSocket!=null) ? new String[]{OAuth2Constants.HTTPS_SCHEME, urlParts[0]} :
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:114]: Any specific reason to change this line?

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java [hunk 7]
```diff
@@ -134,44 +131,41 @@ private static SSLContext getSslContext(BMap<BString, ?> secureSocket) throws Ex
...
-                return buildSslContext(null, tmf.getTrustManagers());
-            }
+            }            
+            tmf = getTrustManagerFactory((BString) cert);
+            return buildSslContext(null, tmf.getTrustManagers());
         }
         if (cert instanceof BMap) {
             BMap<BString, BString> trustStore = (BMap<BString, BString>) cert;
-            if (key != null) {
+            if(key != null){
```
Comment (ThisaruGuruge) [inline native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java:154]: Keep these spaces.

I guess if you ran a build these should be identified by the checkstyle plugin. If it doesn't we might have to check what happened there @MohamedSabthar.

Use this [styling guide](https://raw.githubusercontent.com/wso2/code-quality-tools/v1.4/checkstyle/jdk-17/checkstyle.xml) for your IDE.

```suggestion
            if (key != null) {
```

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java (hunk 6 above)
Comment (ritikk112) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java]: isPresent was a function of Optional container and was removed hence had to change this line too

File: native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java (hunk 7 above)
Comment (ritikk112) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java]: Where can I find the report for the failed build? I want to pinpoint the error but can't find the report, do I have to build and run on my local machine for it?

Comment (MohamedSabthar) [reply native/src/main/java/io/ballerina/stdlib/oauth2/OAuth2Client.java]: Yes, for a detailed report on Checkstyle errors and SpotBugs reports, you should run them locally. Anyway, you should also build and test the project locally before raising the PR. 🙂

**Comment by MohamedSabthar** [comment]
Hi @ritikk112, were you able to build the code locally and apply the fix? If you're facing any issues while building the project, could you please mention them here?

**Comment by MohamedSabthar** [comment]
Closing this PR due to a lack of response from the contributor. Please feel free to reopen or submit a new PR if you wish to continue with this contribution.

**Pull request closed without merging**